File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/infrastructure/listener/UserEventListener.java
package com.github.arhor.aws.graphql.federation.comments.infrastructure.listener;

import com.github.arhor.aws.graphql.federation.comments.service.UserRepresentationService;
import com.github.arhor.aws.graphql.federation.common.event.UserEvent;
import com.github.arhor.aws.graphql.federation.starter.tracing.Trace;
import io.awspring.cloud.sqs.annotation.SqsListener;
import lombok.RequiredArgsConstructor;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.stereotype.Component;

import java.util.UUID;

import static com.github.arhor.aws.graphql.federation.common.constants.AttributesKt.ATTR_IDEMPOTENCY_KEY;
import static com.github.arhor.aws.graphql.federation.common.constants.AttributesKt.ATTR_TRACE_ID;
import static com.github.arhor.aws.graphql.federation.starter.tracing.Utils.withExtendedMDC;

@Trace
@Component
@RequiredArgsConstructor
public class UserEventListener {

    private final UserRepresentationService userRepresentationService;

    @SqsListener("${app-props.events.source.sync-comments-on-user-created-event}")
    public void syncCommentsOnUserCreatedEvent(
        @Payload final UserEvent.Created event,
        @Header(ATTR_TRACE_ID) final UUID traceId,
        @Header(ATTR_IDEMPOTENCY_KEY) final UUID idempotencyKey
    ) {
        withExtendedMDC(traceId, () -> {
            userRepresentationService.createUserRepresentation(
                event.getId(),
                idempotencyKey
            );
        });
    }

    @SqsListener("${app-props.events.source.sync-comments-on-user-deleted-event}")
    public void syncCommentsOnUserDeletedEvent(
        @Payload final UserEvent.Deleted event,
        @Header(ATTR_TRACE_ID) final UUID traceId,
        @Header(ATTR_IDEMPOTENCY_KEY) final UUID idempotencyKey
    ) {
        withExtendedMDC(traceId, () -> {
            userRepresentationService.deleteUserRepresentation(
                event.getId(),
                idempotencyKey
            );
        });
    }
}

File: app-service-votes/src/main/kotlin/com/github/arhor/aws/graphql/federation/votes/api/listener/UserEventListener.kt
package com.github.arhor.aws.graphql.federation.votes.api.listener

import com.github.arhor.aws.graphql.federation.common.constants.ATTR_TRACE_ID
import com.github.arhor.aws.graphql.federation.common.event.UserEvent
import com.github.arhor.aws.graphql.federation.starter.tracing.Trace
import com.github.arhor.aws.graphql.federation.starter.tracing.Utils.withExtendedMDC
import com.github.arhor.aws.graphql.federation.votes.service.UserRepresentationService
import io.awspring.cloud.sqs.annotation.SqsListener
import org.springframework.messaging.handler.annotation.Header
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.stereotype.Component
import java.util.*

@Trace
@Component
class UserEventListener(
    private val userService: UserRepresentationService,
) {

    @SqsListener("\${app-props.events.source.sync-votes-on-user-created-event}")
    fun syncVotesOnUserCreatedEvent(
        @Payload event: UserEvent.Created,
        @Header(ATTR_TRACE_ID) traceId: UUID,
    ) {
        withExtendedMDC(traceId) {
            userService.createUserRepresentation(
                userId = event.id,
                idempotencyKey = traceId,
            )
        }
    }

    @SqsListener("\${app-props.events.source.sync-votes-on-user-deleted-event}")
    fun syncVotesOnVotesDeletedEvent(
        @Payload event: UserEvent.Deleted,
        @Header(ATTR_TRACE_ID) traceId: UUID,
    ) {
        withExtendedMDC(traceId) {
            userService.deleteUserRepresentation(
                userId = event.id,
                idempotencyKey = traceId,
            )
        }
    }
}
